src/blueapi/plugins/data_writing.py

```diff
@@ -0,0 +1,106 @@
+import itertools
+from abc import ABC, abstractmethod, abstractproperty
+from pathlib import Path
+from typing import Optional
+
+import bluesky.plan_stubs as bps
+from aiohttp import ClientSession
+from bluesky.utils import make_decorator
+
+from blueapi.core import MsgGenerator
+from blueapi.plugins.data_writing_server import DataCollection
+
+from .data_writing_server import DataCollection, DataCollectionSetupResult
+
+DATA_COLLECTION_NUMBER = "data_collection_number"
+
+
+class DataCollectionProvider(ABC):
```

Comment:
This should be removed and replaced with <thing Tom is about to push>

Reply:
https://github.com/bluesky/ophyd-async/commit/7b5035067be3d7347fa56a16422e5decdc13433d

Reply:
I can't see a direct replacement to the DataCollectionProvider in the file you've linked there...

your DataCollectionProvider gives a DataCollection, which has 4 things: collection_number, group, raw_data_files_root and nexus_file_path. Tom's work that you've linked has a DirectoryProvider which gives a DirectoryInfo object, which just contains a directory path and filename prefix.

So it's not a direct mapping. Perhaps I'll chat to you about this tomorrow morning in more detail.

Reply:
Cool, got it

Reply:
I've come across a bit of an annoyance; I want to subclass `DirectoryProvider` so it gets the directory from the visit service. However, I want to use aiohttp or some similar asynchronous library... `DirectoryProvider` has one synchronous method, `__call__`.

My solution for now, is to make a subclass `GDADirectoryProvider` (happy to get other suggestions) and have a async `update` method on this. Then the synchronous `__call__` can do something like `asyncio.wait_for` on it.

Reply:
Yep, I think that's what I told @coretl I had in mind at the time

Reply:
As discussed, the pre-processor should do the `async` access to the directory service, then cache the result in `DirectoryProvider`, then the `__call__` method just returns the cached value